src/libtools/linalg.cc

```diff
@@ -19,223 +19,142 @@
-    linalg_eigenvalues( S_eigenvalues, S);
-    if ( S_eigenvalues[0] < 0.0 ) {
-        cerr << " \n Negative eigenvalues in Loewdin transformation " << endl;
-        return -1;
+void linalg_constrained_qrsolve(Eigen::VectorXd &x, Eigen::MatrixXd &A, const Eigen::VectorXd &b,const Eigen::MatrixXd &constr)
+{
+    // check matrix for zero column
+  
+    bool nonzero_found =false;
+    for(int j=0; j<A.cols(); j++) {
+        nonzero_found = A.col(j).isApproxToConstant(0.0,1e-9);
+        if(nonzero_found) {
+            throw std::runtime_error("constrained_qrsolve_zero_column_in_matrix");
+        }
     }
 
-    ub::matrix<double> _diagS = ub::zero_matrix<double>(J.size1(),J.size2() );
-     for ( unsigned _i =0; _i < J.size1() ; _i++){
-
-         _diagS(_i,_i) = 1.0/sqrt(S_eigenvalues[_i]);
-     }
-    ub::matrix<double> _temp = ub::prod( _diagS, ub::trans(S));
-    S = ub::prod( S,_temp );
+    const int NoVariables = x.size();
+    const int NoConstrains = constr.rows(); //number of constraints is number of rows of constr
     
-    _temp = ub::prod(J, S);
-    J = ub::prod( S, _temp);
-    return S_eigenvalues[0];
+    Eigen::HouseholderQR<Eigen::MatrixXd> QR(constr.transpose());
+    Eigen::MatrixXd Q=QR.householderQ();
+ 
+    // Calculate A * Q and store the result in A
+    A = A*Q;
+    // A = [A1 A2], so A2 is just a block of A
+    // [A1 A2] has N rows. A1 has ysize columns 
+    //A2 has 2*ngrid-ysize columns 
+    Eigen::MatrixXd A2 = A.block(0,NoConstrains,A.rows(),NoVariables-NoConstrains);
+    //now perform QR-decomposition of A2 to solve the least-squares problem A2 * z = b
+    //A2 has N rows and (2*ngrid-ysize) columns -> 
+    Eigen::HouseholderQR<Eigen::MatrixXd> QR2(A2);
+    Eigen::VectorXd z=QR2.solve(b);
     
+    // Next two steps assemble vector from y (which is zero-vector) and z
+    Eigen::VectorXd result=Eigen::VectorXd::Zero(NoVariables);
+    for (int i = NoConstrains; i < NoVariables; i++ ) {
+           result[i] = z(i - NoConstrains);
+    }  
+    // To get the final answer this vector should be multiplied by matrix Q
+    x = Q*result;  
+    return;
 }
 
-
-
-int linalg_matrixsqrt(ub::matrix<double> &S){
-    if (S.size1()!=S.size2()) {
-        cerr << " \nMatrix sqrt only works for quadratic matrices " << endl;
-        return -1;
-    }
-  
-    ub::vector<double> S_eigenvalues;
-    linalg_eigenvalues( S_eigenvalues, S);
-    if ( S_eigenvalues[0] < 0.0 ) {
-        cerr << " \n Negative eigenvalues in matrix_sqrt transformation " << endl;
-        return -1;
-    }
-
-    ub::matrix<double> _diagS = ub::zero_matrix<double>(S.size1(),S.size2() );
-     for ( unsigned _i =0; _i < S.size1() ; _i++){
-
-         _diagS(_i,_i) = sqrt(S_eigenvalues[_i]);
-     }
-    ub::matrix<double> _temp = ub::prod( _diagS, ub::trans(S));
-    S = ub::prod( S,_temp );
-    
-  
-    return 1;
-} 
+    bool linalg_eigenvalues(Eigen::MatrixXd&A, Eigen::VectorXd &E, Eigen::MatrixXd&V , int nmax ){
```

Comment:
Can you make nmax const?

Reply:
it is a copy not a reference, so it does not matter